core/src/banking_simulation.rs

```diff
@@ -0,0 +1,900 @@
+        let batch_and_tx_counts = timed_batches_to_send
+            .values()
+            .map(|(_label, batches_with_stats)| {
+                let batches = &batches_with_stats.0;
+                (
+                    batches.len(),
+                    batches.iter().map(|batch| batch.len()).sum::<usize>(),
+                )
+            })
+            .collect::<Vec<_>>();
+        // Convert to a large plain old Vec and drain on it, finally dropping it outside
+        // the simulation loop to avoid jitter due to interleaved deallocs of BTreeMap.
+        let timed_batches_to_send = timed_batches_to_send
+            .into_iter()
+            .map(|(event_time, batches)| {
+                (event_time.duration_since(base_event_time).unwrap(), batches)
+            })
+            .zip_eq(batch_and_tx_counts)
+            .collect::<Vec<_>>();
+
+        let sender_loop = SenderLoop {
+            parent_slot,
+            first_simulated_slot: self.first_simulated_slot,
+            non_vote_sender,
+            tpu_vote_sender,
+            gossip_vote_sender,
+            exit: exit.clone(),
+            raw_base_event_time,
+            base_simulation_time,
+            total_batch_count,
+            timed_batches_to_send,
+        };
+
+        let simulator_loop = SimulatorLoop {
+            bank,
+            parent_slot,
+            first_simulated_slot: self.first_simulated_slot,
+            freeze_time_by_slot,
+            base_event_time,
+            base_simulation_time,
+            poh_recorder,
+            simulated_leader,
+            bank_forks,
+            blockstore,
+            leader_schedule_cache,
+            retransmit_slots_sender,
+            retracer,
+        };
+
+        let simulator_threads = SimulatorThreads {
+            poh_service,
+            banking_stage,
+            broadcast_stage,
+            retracer_thread,
+            exit,
+        };
+
+        (sender_loop, simulator_loop, simulator_threads)
+    }
+
+    pub fn start(
+        self,
+        genesis_config: GenesisConfig,
+        bank_forks: Arc<RwLock<BankForks>>,
+        blockstore: Arc<Blockstore>,
+        block_production_method: BlockProductionMethod,
+    ) -> Result<(), SimulateError> {
+        let (sender_loop, simulator_loop, simulator_threads) = self.prepare_simulation(
+            genesis_config,
+            bank_forks,
+            blockstore,
+            block_production_method,
+        );
+
+        sender_loop.log_starting();
+        // Spawning and entering these two loops must be done at the same time as they're timed.
+        // So, all the mundane setup must be done in advance.
+        let sender_thread = sender_loop.spawn()?;
+        let (sender_thread, retransmit_slots_sender) = simulator_loop.enter(sender_thread);
+
+        simulator_threads.finish(sender_thread, retransmit_slots_sender);
+
+        Ok(())
+    }
```

Comment:
(related to https://github.com/anza-xyz/agave/pull/2733#discussion_r1738929513) so, this is much cleaner, nicely representing high-level workings of `BankingSimulator`, imo. :)

Reply:
well well, this diff is now outdated; had bug... 42162825b2c88459c076dc183b00b475c3e51fc4